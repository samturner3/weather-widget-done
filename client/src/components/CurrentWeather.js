import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import WeatherIcon from 'react-icons-weather';
import ReactLoading from 'react-loading';
import Clock from './Clock';

const Capitalize = (str) => {
	return str.charAt(0).toUpperCase() + str.slice(1);
}

const CurrentWeather = ({
	currentWeatherData,
	weatherLoading,
	placeData,
	country,
	showBack
}) => {
	if (!weatherLoading && currentWeatherData) {
		return (
			<Fragment>
				<div className='widget_left_item-top'>
					<Clock UtcOffSet={placeData.utc_offset} />
					<h1>
						{ Capitalize(currentWeatherData.weather[0].description) }
					</h1>
				</div>
				<div className='widget_left_item-bottom'>
					<div className='current_weather_icon'>
						<div className='current_weather_icon-top'>
							<WeatherIcon name='owm' iconId={currentWeatherData.weather[0].id.toString()} />
						</div>
						<div className='current_weather_icon-bottom'>
							{ Math.round(currentWeatherData.main.temp) }Â°C
						</div>
					</div>
					<div>
						{ placeData.address_components[0].long_name }
						{placeData.address_components[0].long_name !== country && `, ${country}` }
						{
							currentWeatherData.name !== placeData.address_components[0].long_name &&
							`(${currentWeatherData.name})`
						}
					</div>
					<button onClick={e => showBack()}>
						Change Location
					</button>
				</div>
			</Fragment>
		);
	}
	return (
		<ReactLoading type={'spin'} />
	);
};

CurrentWeather.propTypes = {
	currentWeatherData: PropTypes.shape({
		name: PropTypes.string.isRequired,
		weather: PropTypes.arrayOf(PropTypes.shape({
			description: PropTypes.string.isRequired,
			id: PropTypes.number.isRequired
		})).isRequired,
		main: PropTypes.shape({
			temp: PropTypes.number.isRequired
		})
	}),
	weatherLoading: PropTypes.bool.isRequired,
	placeData: PropTypes.shape({
		utc_offset: PropTypes.number.isRequired,
		address_components: PropTypes.arrayOf(PropTypes.shape({
			long_name: PropTypes.string.isRequired
		}))
	}),
	country: PropTypes.string.isRequired,
	showBack: PropTypes.func.isRequired
};

export default CurrentWeather;
