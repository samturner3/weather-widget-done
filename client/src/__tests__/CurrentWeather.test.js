import React from 'react';
import CurrentWeather from '../components/CurrentWeather';

function setup(weatherLoading = false) {
	const props = {
		currentWeatherData: {
			name: 'Melbourne',
			weather: [
				{
					description: 'light rain',
					id: 500
				}
			],
			main: {
				temp: 21.87
			}
		},
		weatherLoading,
		placeData: {
			utc_offset: 600,
			address_components: [
				{
					long_name: 'Melbourne'
				}
			]
		},
		country: 'Australia',
		showBack: jest.fn()
	};
	return {
		wrapper: shallow(<CurrentWeather {...props} />),
		props
	};
}

describe('<Forecast />', () => {
	it('should render the weather info of all five days provided', () => {
		const { wrapper } = setup();
		expect(wrapper).toMatchSnapshot();

		// imported component test
		expect(wrapper.find('TimeClock')).toHaveLength(1);
		expect(wrapper.find('WeatherIcon')).toHaveLength(1);
		expect(wrapper.find('Loading')).toHaveLength(0);
	});

	it('should render the loader when weather data is loading', () => {
		const { wrapper } = setup(true);
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find('Loading')).toHaveLength(1);
	});

	describe('<button/>', () => {
		it('should call showBack() when clicked', () => {
			const { wrapper, props } = setup();
			const button = wrapper.find('button');

			button.simulate('click');

			const showBackSpy = jest.spyOn(props, 'showBack');
			expect(showBackSpy).toHaveBeenCalledTimes(1);
		});
	});
});
