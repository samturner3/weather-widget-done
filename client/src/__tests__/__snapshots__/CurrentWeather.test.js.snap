// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Forecast /> should render the loader when weather data is loading 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CurrentWeather
    country="Australia"
    currentWeatherData={
        Object {
            "main": Object {
              "temp": 21.87,
            },
            "name": "Melbourne",
            "weather": Array [
              Object {
                "description": "light rain",
                "id": 500,
              },
            ],
          }
    }
    placeData={
        Object {
            "address_components": Array [
              Object {
                "long_name": "Melbourne",
              },
            ],
            "utc_offset": 600,
          }
    }
    showBack={[Function]}
    weatherLoading={true}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "color": "#fff",
      "delay": 0,
      "height": 64,
      "type": "spin",
      "width": 64,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "color": "#fff",
        "delay": 0,
        "height": 64,
        "type": "spin",
        "width": 64,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`<Forecast /> should render the weather info of all five days provided 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CurrentWeather
    country="Australia"
    currentWeatherData={
        Object {
            "main": Object {
              "temp": 21.87,
            },
            "name": "Melbourne",
            "weather": Array [
              Object {
                "description": "light rain",
                "id": 500,
              },
            ],
          }
    }
    placeData={
        Object {
            "address_components": Array [
              Object {
                "long_name": "Melbourne",
              },
            ],
            "utc_offset": 600,
          }
    }
    showBack={[Function]}
    weatherLoading={false}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <div
          className="widget_left_item-top"
>
          <TimeClock
                    UtcOffSet={600}
          />
          <h1>
                    Light rain
          </h1>
</div>,
        <div
          className="widget_left_item-bottom"
>
          <div
                    className="current_weather_icon"
          >
                    <div
                              className="current_weather_icon-top"
                    >
                              <WeatherIcon
                                        iconId="500"
                                        name="owm"
                              />
                    </div>
                    <div
                              className="current_weather_icon-bottom"
                    >
                              22
                              °C
                    </div>
          </div>
          <div>
                    Melbourne
                    , Australia
          </div>
          <button
                    onClick={[Function]}
          >
                    Change Location
          </button>
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <TimeClock
              UtcOffSet={600}
/>,
            <h1>
              Light rain
</h1>,
          ],
          "className": "widget_left_item-top",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "UtcOffSet": 600,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Light rain",
            },
            "ref": null,
            "rendered": "Light rain",
            "type": "h1",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="current_weather_icon"
>
              <div
                            className="current_weather_icon-top"
              >
                            <WeatherIcon
                                          iconId="500"
                                          name="owm"
                            />
              </div>
              <div
                            className="current_weather_icon-bottom"
              >
                            22
                            °C
              </div>
</div>,
            <div>
              Melbourne
              , Australia
</div>,
            <button
              onClick={[Function]}
>
              Change Location
</button>,
          ],
          "className": "widget_left_item-bottom",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="current_weather_icon-top"
>
                  <WeatherIcon
                                    iconId="500"
                                    name="owm"
                  />
</div>,
                <div
                  className="current_weather_icon-bottom"
>
                  22
                  °C
</div>,
              ],
              "className": "current_weather_icon",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <WeatherIcon
                    iconId="500"
                    name="owm"
/>,
                  "className": "current_weather_icon-top",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "iconId": "500",
                    "name": "owm",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    22,
                    "°C",
                  ],
                  "className": "current_weather_icon-bottom",
                },
                "ref": null,
                "rendered": Array [
                  22,
                  "°C",
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Melbourne",
                ", Australia",
                false,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Melbourne",
              ", Australia",
              false,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Change Location",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Change Location",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <div
            className="widget_left_item-top"
>
            <TimeClock
                        UtcOffSet={600}
            />
            <h1>
                        Light rain
            </h1>
</div>,
          <div
            className="widget_left_item-bottom"
>
            <div
                        className="current_weather_icon"
            >
                        <div
                                    className="current_weather_icon-top"
                        >
                                    <WeatherIcon
                                                iconId="500"
                                                name="owm"
                                    />
                        </div>
                        <div
                                    className="current_weather_icon-bottom"
                        >
                                    22
                                    °C
                        </div>
            </div>
            <div>
                        Melbourne
                        , Australia
            </div>
            <button
                        onClick={[Function]}
            >
                        Change Location
            </button>
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <TimeClock
                UtcOffSet={600}
/>,
              <h1>
                Light rain
</h1>,
            ],
            "className": "widget_left_item-top",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "UtcOffSet": 600,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Light rain",
              },
              "ref": null,
              "rendered": "Light rain",
              "type": "h1",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="current_weather_icon"
>
                <div
                                className="current_weather_icon-top"
                >
                                <WeatherIcon
                                                iconId="500"
                                                name="owm"
                                />
                </div>
                <div
                                className="current_weather_icon-bottom"
                >
                                22
                                °C
                </div>
</div>,
              <div>
                Melbourne
                , Australia
</div>,
              <button
                onClick={[Function]}
>
                Change Location
</button>,
            ],
            "className": "widget_left_item-bottom",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <div
                    className="current_weather_icon-top"
>
                    <WeatherIcon
                                        iconId="500"
                                        name="owm"
                    />
</div>,
                  <div
                    className="current_weather_icon-bottom"
>
                    22
                    °C
</div>,
                ],
                "className": "current_weather_icon",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <WeatherIcon
                      iconId="500"
                      name="owm"
/>,
                    "className": "current_weather_icon-top",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "iconId": "500",
                      "name": "owm",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "type": "div",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      22,
                      "°C",
                    ],
                    "className": "current_weather_icon-bottom",
                  },
                  "ref": null,
                  "rendered": Array [
                    22,
                    "°C",
                  ],
                  "type": "div",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Melbourne",
                  ", Australia",
                  false,
                ],
              },
              "ref": null,
              "rendered": Array [
                "Melbourne",
                ", Australia",
                false,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Change Location",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Change Location",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
